src/main/java/com/overops/plugins/service/impl/ReportBuilder.java

```diff
@@ -186,290 +189,198 @@ private Version createVersion(EventResult e) {
-		protected List<OOReportEvent> topEvents;
-		protected Collection<EventResult> filter;
-		
-	}
-	
-	private static boolean allowEvent(EventResult event, Pattern pattern) {
-		
-		if (pattern == null ) {
-			return true;
+
+	public static ReportBuilder.QualityReport execute(ApiClient apiClient, QueryOverOps properties, ReliabilityReport reliabilityReport,
+													  ReliabilityReportInput input, DateTime activeWindowStart, boolean runRegression, PrintStream output) {
+		boolean countGate = false;
+		if (properties.getMaxErrorVolume() != 0 || properties.getMaxUniqueErrors() != 0) {
+			countGate = true;
 		}
-		
-		String json = new Gson().toJson(event);
-		boolean result = !pattern.matcher(json).find();
-		
-		return result;
-	}
-	
-	private static List<EventResult> getSortedEventsByVolume(Collection<EventResult> events) {
-		
-		List<EventResult> result = new ArrayList<EventResult>(events);
-		
-		result.sort((o1, o2) -> {
-			long v1;
-			long v2;
-
-			if (o1.stats != null) {
-				v1 = o1.stats.hits;
-			} else {
-				v1 = 0;
-			}
+		boolean checkMaxEventGate = properties.getMaxErrorVolume() != 0;
+		boolean checkUniqueEventGate = properties.getMaxUniqueErrors() != 0;
 
-			if (o2.stats != null) {
-				v2 = o2.stats.hits;
-			} else {
-				v2 = 0;
-			}
+		ReliabilityReport.ReliabilityReportItem report = reliabilityReport.items.values().iterator().next();
+		List<SeriesRow> rows = new ArrayList<>();
+		if (report.errors != null) {
+			rows = report.errors.readRows();
+		}
 
-			return (int)(v2 - v1);
-		});
-		
-		return result;
-	}
-	
-	private static void addEvent(Set<EventResult> events, EventResult event, 
-		Pattern pattern, PrintStream output, boolean verbose) {
-		
-		if (allowEvent(event, pattern)) {
-			events.add(event);
-		} else if ((output != null) && (verbose)) {
-			output.println(event + " did not match regexFilter and was skipped");
+		if (report.failures != null) {
+			rows.addAll(report.failures.readRows());
 		}
-	}
-	
-	private static Collection<EventResult> filterEvents(RateRegression rateRegression,
-			Pattern pattern, PrintStream output, boolean verbose) {
-		
-		Set<EventResult> result = new HashSet<>();
-		
-		if (pattern != null) {
-		
-			for (EventResult event : rateRegression.getNonRegressions()) {
-				addEvent(result, event, pattern, output, verbose);
-			}
-			
-			for (EventResult event : rateRegression.getAllNewEvents().values()) {
-				addEvent(result, event, pattern, output, verbose);
+		List<EventRow> errors = filterEvents(rows, properties.getRegexFilter());
+		errors.sort((o1, o2) -> (int) o2.hits - (int) o1.hits);
+
+		List<OOReportEvent> topErrors = new ArrayList<>();
+		List<OOReportEvent> resurfacedErrors = new ArrayList<>();
+		List<OOReportEvent> criticalErrors = new ArrayList<>();
+
+		for(EventRow row : errors) {
+			if (countGate && topErrors.size() <= properties.getPrintTopIssues()) {
+				String link = Optional.ofNullable(row.id).map(id -> id.split(",")[0])
```

Comment:
perhaps we should break this to a method and have more line breaks with some inline comments as it's a bit hard to read it